tox.ini

```diff
@@ -77,36 +78,15 @@ deps =
     pytest-cov
 
 [testenv:func]
-setenv =
-  TEST_JUJU_CHANNEL = 2.9/stable
 changedir = {toxinidir}/tests/functional
-commands = functest-run-suite
+commands = functest-run-suite {posargs:--keep-faulty-model}
 deps =
   git+https://github.com/openstack-charmers/zaza.git#egg=zaza
   -r {toxinidir}/tests/functional/requirements.txt
 
-[testenv:func33-jammy]
-setenv =
-  TEST_JUJU_CHANNEL = 3.3/stable
+[testenv:func3]
 changedir = {toxinidir}/tests/functional
-commands =
-  # NOTE(mertkirpici): functest-run-suite -b jammy focal
-  # or functest-run-suite -b jammy -b focal does not work
-  # https://github.com/openstack-charmers/zaza/issues/614
-  functest-run-suite -b jammy
-deps =
-  git+https://github.com/openstack-charmers/zaza.git@libjuju-3.1#egg=zaza[juju-33]
-  -r {toxinidir}/tests/functional/requirements.txt
-
-[testenv:func33-focal]
-setenv =
-  TEST_JUJU_CHANNEL = 3.3/stable
-changedir = {toxinidir}/tests/functional
-commands =
-  # NOTE(mertkirpici): functest-run-suite -b jammy focal
-  # or functest-run-suite -b jammy -b focal does not work
-  # https://github.com/openstack-charmers/zaza/issues/614
-  functest-run-suite -b focal
+commands = functest-run-suite {posargs:--keep-faulty-model}
 deps =
-  git+https://github.com/openstack-charmers/zaza.git@libjuju-3.1#egg=zaza[juju-33]
+  git+https://github.com/openstack-charmers/zaza.git@libjuju-3.1#egg=zaza
```

Comment:
I think we should use the constraints instead of old branch, which does not seems to have any updates.

I tested it and we can use:
```bash
TEST_JUJU3=1 pip install git+https://github.com/openstack-charmers/zaza.git#egg=zaza --constraint https://raw.githubusercontent.com/openstack-charmers/zaza/master/constraints-juju31.txt
# or
TEST_JUJU3=1 pip install git+https://github.com/openstack-charmers/zaza.git#egg=zaza --constraint https://raw.githubusercontent.com/openstack-charmers/zaza/master/constraints-juju32.txt
# or 
TEST_JUJU3=1 pip install git+https://github.com/openstack-charmers/zaza.git#egg=zaza --constraint https://raw.githubusercontent.com/openstack-charmers/zaza/master/constraints-juju33.txt
```

Reply:
I think there is a conflict with python-libjuju and zaza. python-libjuju maintainers mention about we can use the latest version for all 3.x juju. This make zaza's constraint looks unnecessary.

Overall I think it's ok we use the latest version that zaza supported, which is juju33. Otherwise we need to create our own constraints.

Reply:
Also I have question of your example: 

```
TEST_JUJU3=1 pip install git+https://github.com/openstack-charmers/zaza.git#egg=zaza --constraint https://raw.githubusercontent.com/openstack-charmers/zaza/master/constraints-juju31.txt
```

Where you want to put it? I test it on tox and it's not working. The juju version is still 3.1.2

Reply:
Please check any repo I did support for Juju 3.4.